honeybadgermpc/reed_solomon.py

```diff
@@ -313,37 +311,35 @@ def _robust_update(self):
                 break
 
             num_agreement = len(self._available_points) - len(errors)
-            if num_agreement >= self._min_points_required():
-                # Errors detected considered to be confirmed errors
-                self._confirmed_errors |= set(errors)
-                self._available_points -= set(errors)
-                error_indices = []
-                for e in errors:
-                    error_idx = self._z.index(e)
-                    error_indices.append(error_idx)
-                    del self._z[error_idx]
-                self._num_decoded += 1
-                self._available_data = self._available_data[1:]
-                self._partial_result.append(decoded)
-
-                for error_idx in error_indices:
-                    for i in range(len(self._available_data)):
-                        del self._available_data[i][error_idx]
-            else:
+            if num_agreement < self._min_points_required():
                 break
 
+            self._num_decoded += 1
+            self._available_data = self._available_data[1:]
+            self._partial_result.append(decoded)
+
+            # Errors detected considered to be confirmed errors
+            self._confirmed_errors |= set(errors)
+            self._available_points -= set(errors)
+
+            for e in errors:
+                error_idx = self._z.index(e)
+
+                del self._z[error_idx]
+                for i in range(len(self._available_data)):
+                    del self._available_data[i][error_idx]
+
         # We're done
         if self._num_decoded == self.batch_size:
             self._result = self._partial_result
 
     # Public API
     def add(self, idx, data):
-        if self.done() or idx in self._available_points or \
-                idx in self._confirmed_errors:
+        if self.done() or idx in [*self._available_points, *self._confirmed_errors]:
```

Comment:
similar comment here, this will create a new list which takes up additional memory..

Reply:
This will take up `4*len(self._available_points) + 4*len(self._confirmed_errors) + 22` additional bytes of memory I believe. It'll then be garbage collected during the next collection loop, as the refcount of the created list should immediately hit 0 after the if statement. I think the overhead is worth the cleaner code, but I'll make the change.

Reply:
that link also is making me think that numpy could be a really good idea in our critical path if we can manage to do it cleanly.

Reply:
[Did some more reading](https://rushter.com/blog/python-memory-managment/), this will likely allocate the array in an already allocated `arena` (fairly quick, as is deallocation), so an actual os allocation is unlikely with how aggressive python memory caching is. This is likely a safe pattern outside of the case where the array created is greater than 4kb (i.e. < 500 elements), unless there's existing memory pressure.

Reply:
This pattern when using a tuple rather than a list is much safer, as it will deallocate the tuple immediately after the if statement, as tuples use the reference-count based GC vs. the cyclic cycle detecting algorithm used for lists which only runs periodically.